File: src/Systems/Objective-C/patterns/mediator.m
#import <Foundation/Foundation.h>

@protocol MediatorColleague <NSObject>
- (NSString *)receiveFrom:(NSString *)sender message:(NSString *)message;
@end

@interface CheckoutMediator : NSObject {
    NSMutableDictionary *_colleagues;
}
- (void)registerColleague:(id<MediatorColleague>)colleague named:(NSString *)name;
- (NSString *)sendFrom:(NSString *)sender to:(NSString *)recipient message:(NSString *)message;
@end

@implementation CheckoutMediator
- (instancetype)init {
    if ((self = [super init])) {
        _colleagues = [NSMutableDictionary dictionary];
    }
    return self;
}

- (void)registerColleague:(id<MediatorColleague>)colleague named:(NSString *)name {
    [_colleagues setObject:colleague forKey:name];
}

- (NSString *)sendFrom:(NSString *)sender to:(NSString *)recipient message:(NSString *)message {
    id<MediatorColleague> colleague = [_colleagues objectForKey:recipient];
    if (colleague == nil) {
        [NSException raise:@"UnknownColleague"
                    format:@"unknown colleague: %@", recipient];
    }
    return [colleague receiveFrom:sender message:message];
}
@end

@interface PaymentColleague : NSObject <MediatorColleague>
@end

@implementation PaymentColleague
- (NSString *)receiveFrom:(NSString *)sender message:(NSString *)message {
    return [NSString stringWithFormat:@"payment<-%@:%@", sender, message];
}
@end

@interface InventoryColleague : NSObject <MediatorColleague>
@end

@implementation InventoryColleague
- (NSString *)receiveFrom:(NSString *)sender message:(NSString *)message {
    return [NSString stringWithFormat:@"inventory<-%@:%@", sender, message];
}
@end

static BOOL verifyMediator(void) {
    CheckoutMediator *mediator = [CheckoutMediator new];
    [mediator registerColleague:[PaymentColleague new] named:@"payment"];
    [mediator registerColleague:[InventoryColleague new] named:@"inventory"];

    NSString *paid = [mediator sendFrom:@"payment" to:@"inventory" message:@"paid"];
    NSString *reserved = [mediator sendFrom:@"inventory" to:@"payment" message:@"reserved"];

    BOOL rejectedUnknown = NO;
    @try {
        (void)[mediator sendFrom:@"payment" to:@"shipping" message:@"dispatch"];
    }
    @catch (NSException *exception) {
        rejectedUnknown = [[exception name] isEqualToString:@"UnknownColleague"] &&
            [[exception reason] isEqualToString:@"unknown colleague: shipping"];
    }

    return [paid isEqualToString:@"inventory<-payment:paid"] &&
        [reserved isEqualToString:@"payment<-inventory:reserved"] &&
        rejectedUnknown;
}

int main(void) {
    @autoreleasepool {
        NSCAssert(verifyMediator(), @"Objective-C Mediator contract failed");
        puts("Objective-C Mediator: passed");
    }
    return 0;
}

File: src/Scripting/Perl/mediator.pl
use strict;
use warnings;

my %receivers;

sub register_colleague {
    my ($name, $receiver) = @_;
    die "receiver must be a code reference\n" unless ref($receiver) eq 'CODE';
    $receivers{$name} = $receiver;
}

sub send_message {
    my ($sender, $recipient, $message) = @_;
    my $receiver = $receivers{$recipient};
    die "UnknownColleague:$recipient\n" unless defined $receiver;
    return $receiver->($sender, $message);
}

my @payment_events;
my @inventory_events;

register_colleague('payment', sub {
    my ($sender, $message) = @_;
    push @payment_events, "$sender:$message";
    return 'payment-ack';
});

register_colleague('inventory', sub {
    my ($sender, $message) = @_;
    push @inventory_events, "$sender:$message";
    return 'inventory-ack';
});

my $inventory_ack = send_message('payment', 'inventory', 'paid');
my $payment_ack = send_message('inventory', 'payment', 'reserved');

die "inventory routing failed\n"
    unless $inventory_ack eq 'inventory-ack'
    && join(',', @inventory_events) eq 'payment:paid';

die "payment routing failed\n"
    unless $payment_ack eq 'payment-ack'
    && join(',', @payment_events) eq 'inventory:reserved';

my $unknown_error = '';
eval { send_message('payment', 'shipping', 'dispatch'); 1 } or $unknown_error = $@;
die "missing unknown colleague failure\n"
    unless $unknown_error =~ /UnknownColleague:shipping/;

print "Perl Mediator: passed\n";
